File: .leetcode/264.丑数-ii.java
import java.util.HashSet;
import java.util.PriorityQueue;

/*
 * @lc app=leetcode.cn id=264 lang=java
 *
 * [264] 丑数 II
 */

// @lc code=start
class Solution {
    public int nthUglyNumber(int n) {
        HashSet<Long> s = new HashSet<Long>();
        PriorityQueue<Long> heap = new PriorityQueue<Long>();
        s.add(1L);
        heap.add(1L);
        int[]factor = {2,3,5};
        long temp = 1L;
        for(int i = 1; i <= n;i++){
            temp = heap.poll();
            for(int j = 0;j < 3;j++){
                if(s.add(temp * factor[j])){
                    heap.add(temp * factor[j]);
                }
            }
        }
        return (int)temp;
    }
}
// @lc code=end

File: .leetcode/264.丑数-ii.cpp
/*
 * @lc app=leetcode.cn id=264 lang=cpp
 *
 * [264] 丑数 II
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int nthUglyNumber(int n) {
        if(n == 1)return 1;
        priority_queue<int,vector<int>,greater<int>>q;
        q.push(2);
        q.push(3);
        q.push(5);
        int index = -1;
        int i = 2;
        while(true){
            index++;
            int temp = q.top();
            q.pop();
            if(index == n)return temp; 
            if(i <=563){
                q.push(temp*2);
                q.push(temp*3);
                q.push(temp*5);
            }
        }
       return index;
    }
};
// @lc code=end

